### ISSUE TITLE
pytorch scripts/run_pretraining.sh calculates sequences/sec incorrectly

### ISSUE THREAD

**Issue by David-Levinthal** [NVIDIA/DeepLearningExamples#242]
this is a very minor issue...but a bit confusing   :-)
the batch size used is actually args.train_batch_size/ gradient_accumulation steps
line 250 of run_pretraining.py
args.train_batch_size = args.train_batch_size // args.gradient_accumulation_steps

this is clear fromt he printout in main
            logger.info("***** Running training *****")
            # logger.info("  Num examples = %d", len(train_data))
            logger.info("  Batch size = %d", args.train_batch_size)

thus in run_pretraining.sh  the following:
train_perf=$(awk 'BEGIN {print ('$throughput' * '$num_gpus' * '$train_batch_size')}')
should read
train_perf=$(awk 'BEGIN {print ('$throughput' * '$num_gpus' * '$train_batch_size' / '$gradient_accumulation_steps')}')

**Comment by sharathts** [comment]
Thanks for pointing this. Will update it shortly.

**Comment by sharathts** [comment]
Being fixed https://github.com/NVIDIA/DeepLearningExamples/pull/259

**Comment by David-Levinthal** [comment]
I would actually recommend getting rid of the tqdm call and using explicit
calls to time to get the time/step and thus the iterations/sec or
sequences/sec
this would
reduce overhead from all the print executions
simplify post processing a LOT
reduce the size of the log file by 2 orders of magnitude

**Comment by sharathts** [comment]
The total steps logged is also off by a factor of gradient accumulation steps. Both the fixes have been pushed to the PR above. Thanks!

**Comment by David-Levinthal** [comment]
can you get rid of tqdm...and just use calls to time around each step? that
really would simplify everything

**Comment by David-Levinthal** [comment]
With the fixes made to date the throughput printed out will still have issues as it does not grab the last of the tqdm outputs for the step. this can be checked by comparing the throughput printed out with the throughput calculated by manually getting the last tqdm output or even dividing the total number of batches/gpu by the total time printed out at the end of phase 1  :-)
consider the following output taken from a short phase1 run that stopped after 200 steps with an aggregation size of 128 on 4 GPUs..note this used a run_pretraining script from before you added the division by the aggregation size

Step:198 Average Loss = 7.992095947265625 Step Loss = 7.76953125 LR 0.006
Iteration:  32%|\u2588\u2588\u2588\u258f      | 3588/11189 [18:51<39:46,  3.19it/s]Step 199 LR 0.00042426406871192866
Step:199 Average Loss = 7.99322509765625 Step Loss = 7.96875 LR 0.006
Iteration:  33%|\u2588\u2588\u2588\u258e      | 3716/11189 [19:32<39:00,  3.19it/s]Step 200 LR 0.0
10/10/2019 22:33:50 - INFO - __main__ -   Total Steps:26240 Final Loss = 7.990135192871094
10/10/2019 22:33:50 - INFO - __main__ -   ** ** * Saving fine - tuned model ** ** * 
Total time taken 8335.677252531052
 
*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
*****************************************
+ set +x
finished pretraining, starting benchmarking
training throughput phase1: 89129 sequences/second
average loss: 7.99322509765625
final loss: 7.990135192871094

If you take the total time of 8335.68 sec…and that time ran 200 steps of 128 batches
200*128/8335.68 = 3.07 it/sec

but if you use the 89129 seq/sec throughput correct for the aggregation size and divide by 4 for the number of GPUs and by the batch size to get the iterations/sec the calculation used you get
89129/(128*4*64) = 2.72
which is roughly the first of the tqdm values, rather than the last ones (~ 3.19)
I uploaded the log file
[bert_bs64_np4_128_def.log](https://github.com/NVIDIA/DeepLearningExamples/files/3753182/bert_bs64_np4_128_def.log)

**Comment by sharathts** [comment]
https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/LanguageModeling/BERT/scripts/run_pretraining.sh#L151

The above line always picks up the latest tqdm output since `tail -1`

Also, I wasn't able to reproduce 

> training throughput phase1: 89129 sequences/second

With the logfile attached above, I got the following output which is the correct performance.

>  training throughput phase1: 816.64 sequences/second
> average loss: 7.99322509765625
> final loss: 7.990135192871094

Have you changed the calculation here? https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/LanguageModeling/BERT/scripts/run_pretraining.sh#L155

Note: You are running train_batch_size=8192 and gradient_accumulation_steps=128 

sequences are computed in the above link as follows:
3.19 * 4 * 8192 / 128 = 816.64 

>If you take the total time of 8335.68 sec…and that time ran 200 steps of 128 batches
>200*128/8335.68 = 3.07 it/sec

3.07it/sec is slightly lower than 3.19it/sec since 8335.68 seconds includes model setup + data loading + training. Training alone progresses at 3.19it/sec
3.19

**Comment by sharathts** [comment]
Your suggestion to remove tqdm has been duly noted. It should be updated in a subsequent release. :)

**Comment by David-Levinthal** [comment]
Just to illustrate the tqdm issue a bit more clearly (ie that the script grabs the first value and not the last
The following is from the end of a (different) full run of phase1 (sequence length 128). The accumulation size was 64 and the local batch size was 64
The final throughput summary of the run was
training throughput phase1: 2723.84 sequences/second
we can use this, the accumulation size and number of nodes to compute an average rate of iterations/sec that the base node would print out. This yields:
2724/(16*64) = 2.66 it/sec

We can process the log to show tqdms’ last value (of 64 printed per step)
cat bert_16gpu_phase1_pyt.log | tail -n 25
Iteration:   6%|\u258c         | 661/11337 [03:37<57:35,  3.09it/s]Step 7032 LR 0.00017518714874213283
Step:7032 Average Loss = 1.4574737548828125 Step Loss = 1.6171875 LR 0.006
Iteration:   6%|\u258b         | 725/11337 [03:57<57:13,  3.09it/s]Step 7033 LR 0.00015992325525180656
Step:7033 Average Loss = 1.443267822265625 Step Loss = 1.4541015625 LR 0.006
Iteration:   7%|\u258b         | 789/11337 [04:18<56:53,  3.09it/s]Step 7034 LR 0.0001430397079704331
Step:7034 Average Loss = 1.4429473876953125 Step Loss = 1.462890625 LR 0.006
Iteration:   8%|\u258a         | 853/11337 [04:39<56:36,  3.09it/s]Step 7035 LR 0.0001238760208522985
Step:7035 Average Loss = 1.449462890625 Step Loss = 1.4208984375 LR 0.006
Iteration:   8%|\u258a         | 917/11337 [05:00<56:11,  3.09it/s]Step 7036 LR 0.00010114434748482682
Step:7036 Average Loss = 1.446441650390625 Step Loss = 1.435546875 LR 0.006
Iteration:   9%|\u258a         | 981/11337 [05:21<55:52,  3.09it/s]Step 7037 LR 7.151985398522354e-05
Step:7037 Average Loss = 1.46856689453125 Step Loss = 1.4853515625 LR 0.006
Iteration:   9%|\u2589         | 1045/11337 [05:42<55:32,  3.09it/s]Step 7038 LR 0.0
11/13/2019 19:47:17 - INFO - __main__ -   Total Steps:7048.0 Final Loss = 1.4431438446044922
11/13/2019 19:47:17 - INFO - __main__ -   ** ** * Saving fine - tuned model ** ** * 
Total time taken 149858.59364962578
 
*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
*****************************************
+ set +x
finished pretraining, starting benchmarking
training throughput phase1: 2723.84 sequences/second
average loss: 1.46856689453125
final loss: 1.4431438446044922
 
On the other hand, we can extract tqdms’ first value and one might ask
which one is more likely to yield 2724 seq/sec?   😊
 
$ cat bert_16gpu_phase1_pyt.log | tail -n 25 | cut -f1 -d]
Iteration:   5%|\u258c         | 598/11337 [03:16<1:03:39,  2.81it/s
Step:7032 Average Loss = 1.4574737548828125 Step Loss = 1.6171875 LR 0.006
Iteration:   6%|\u258c         | 662/11337 [03:37<1:05:44,  2.71it/s
Step:7033 Average Loss = 1.443267822265625 Step Loss = 1.4541015625 LR 0.006
Iteration:   6%|\u258b         | 726/11337 [03:58<1:06:21,  2.66it/s
Step:7034 Average Loss = 1.4429473876953125 Step Loss = 1.462890625 LR 0.006
Iteration:   7%|\u258b         | 790/11337 [04:19<1:04:03,  2.74it/s
Step:7035 Average Loss = 1.449462890625 Step Loss = 1.4208984375 LR 0.006
Iteration:   8%|\u258a         | 854/11337 [04:40<1:01:47,  2.83it/s
Step:7036 Average Loss = 1.446441650390625 Step Loss = 1.435546875 LR 0.006
Iteration:   8%|\u258a         | 918/11337 [05:01<1:04:16,  2.70it/s
Step:7037 Average Loss = 1.46856689453125 Step Loss = 1.4853515625 LR 0.006
Iteration:   9%|\u258a         | 982/11337 [05:21<1:04:48,  2.66it/s
11/13/2019 19:47:17 - INFO - __main__ -   Total Steps:7048.0 Final Loss = 1.4431438446044922
11/13/2019 19:47:17 - INFO - __main__ -   ** ** * Saving fine - tuned model ** ** * 
Total time taken 149858.59364962578
 
*****************************************
Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
*****************************************
+ set +x
finished pretraining, starting benchmarking
training throughput phase1: 2723.84 sequences/second
average loss: 1.46856689453125
final loss: 1.4431438446044922

Conclusion: using tqdm in production code might not be the best choice.
